src/mysql_k8s_helpers.py

```diff
@@ -335,16 +451,284 @@ def execute_backup_commands(
+            logger.exception("Failed to retrieve backup", exc_info=e)
+            logger.error(f"Stdout: {e.stdout}")
+            logger.error(f"Stderr: {e.stderr}")
+            raise MySQLRetrieveBackupWithXBCloudError(e.stderr)
+        except Exception as e:
+            logger.exception("Failed to retrieve backup", exc_info=e)
+            raise MySQLRetrieveBackupWithXBCloudError(e)
+
+    def prepare_backup_for_restore(self, backup_location: str) -> Tuple[str, str]:
+        """Prepare the backup in the provided dir for restore."""
+        try:
+            innodb_buffer_pool_size, _ = self.get_innodb_buffer_pool_parameters()
+        except MySQLGetInnoDBBufferPoolParametersError as e:
+            raise MySQLPrepareBackupForRestoreError(e)
+
+        prepare_backup_command = f"""
+xtrabackup --prepare
+        --use-memory={innodb_buffer_pool_size}
+        --no-version-check
+        --rollback-prepared-trx
+        --xtrabackup-plugin-dir=/usr/lib64/xtrabackup/plugin
+        --target-dir={backup_location}
+""".split()
+
+        try:
+            process = self.container.exec(
+                prepare_backup_command,
+                user="mysql",
+                group="mysql",
+            )
+            stdout, stderr = process.wait_output()
+            return (stdout, stderr)
+        except ExecError as e:
+            logger.exception("Failed to prepare backup for restore", exc_info=e)
+            logger.error(f"Stdout: {e.stdout}")
+            logger.error(f"Stderr: {e.stderr}")
+            raise MySQLPrepareBackupForRestoreError(e.stderr)
+        except Exception as e:
+            logger.exception("Failed to prepare backup for restore", exc_info=e)
+            raise MySQLPrepareBackupForRestoreError(e)
+
+    def empty_data_files(self) -> None:
+        """Empty the mysql data directory in preparation of backup restore."""
+        empty_data_files_command = "find /var/lib/mysql/ -not -path /var/lib/mysql/mysql_sst_* -not -path /var/lib/mysql/ -delete".split()
+
+        try:
+            process = self.container.exec(
+                empty_data_files_command,
+                user="mysql",
+                group="mysql",
+            )
+            process.wait_output()
+        except ExecError as e:
+            logger.exception(
+                "Failed to empty data directory in prep for backup restore", exc_info=e
+            )
+            logger.error(f"Stdout: {e.stdout}")
+            logger.error(f"Stderr: {e.stderr}")
+            raise MySQLEmptyDataDirectoryError(e.stderr)
+        except Exception as e:
+            logger.exception(
+                "Failed to empty data directory in prep for backup restore", exc_info=e
+            )
+            raise MySQLEmptyDataDirectoryError(e)
+
+    def restore_backup(self, backup_location: str) -> Tuple[str, str]:
+        """Restore the provided prepared backup."""
+        restore_backup_command = f"""
+xtrabackup --defaults-file=/etc/mysql
```

Comment:
are options from MySQL config applied during restoration?
I think `--defaults-file` can read file only, not a dir

Reply:
you were right: `--defaults-file` was not picking up the MySQL config during restoration.

--defaults-file=/etc/mysql
```
unit-mysql-k8s4-0: 14:45:24 INFO unit.mysql-k8s4/0.juju-log Restoring the backup                                                                                                                           
unit-mysql-k8s4-0: 14:45:25 DEBUG unit.mysql-k8s4/0.juju-log Stdout of xtrabackup move-back command:                                                                                                       
unit-mysql-k8s4-0: 14:45:25 DEBUG unit.mysql-k8s4/0.juju-log Stderr of xtrabackup move-back command: 2023-02-15T19:45:25.034827-00:00 0 [Note] [MY-011825] [Xtrabackup] recognized server arguments: --defa
ults_group=mysqld --datadir=/var/lib/mysql                                                                                                                                                                 
2023-02-15T19:45:25.034873-00:00 0 [Note] [MY-011825] [Xtrabackup] recognized client arguments: --no-version-check=1 --move-back=1 --force-non-empty-directories=1 --xtrabackup-plugin-dir=/usr/lib64/xtrab
ackup/plugin --target-dir=/var/lib/mysql/mysql_sst_aQOq 
```

--defaults-file=/etc/mysql/my.cnf
```
unit-mysql-k8s4-0: 14:45:24 INFO unit.mysql-k8s4/0.juju-log Restoring the backup                                                                                                                           
unit-mysql-k8s4-0: 14:45:25 DEBUG unit.mysql-k8s4/0.juju-log Stdout of xtrabackup move-back command:       
unit-mysql-k8s4-0: 14:46:27 DEBUG unit.mysql-k8s4/0.juju-log Stderr of xtrabackup move-back command: 2023-02-15T19:46:27.048630-00:00 0 [Note] [MY-011825] [Xtrabackup] recognized server arguments: --data
dir=/var/lib/mysql --innodb_buffer_pool_size=12482248704 --defaults_group=mysqld --datadir=/var/lib/mysql                                                                                                  
2023-02-15T19:46:27.048725-00:00 0 [Note] [MY-011825] [Xtrabackup] recognized client arguments: --no-version-check=1 --move-back=1 --force-non-empty-directories=1 --xtrabackup-plugin-dir=/usr/lib64/xtrab
ackup/plugin --target-dir=/var/lib/mysql/mysql_sst_IdDH
```